// @tldraw/state/dist-esm/lib/core/index.mjs
import { singleton } from "./helpers.mjs";
import { ArraySet } from "./ArraySet.mjs";
import { atom, isAtom } from "./Atom.mjs";
import { UNINITIALIZED, computed, getComputedInstance, isUninitialized, withDiff } from "./Computed.mjs";
import { EffectScheduler, react, reactor } from "./EffectScheduler.mjs";
import { unsafe__withoutCapture, whyAmIRunning } from "./capture.mjs";
import { EMPTY_ARRAY } from "./helpers.mjs";
import { isSignal } from "./isSignal.mjs";
import { transact, transaction } from "./transactions.mjs";
import { RESET_VALUE } from "./types.mjs";
const currentApiVersion = 1;
const actualApiVersion = singleton("apiVersion", () => currentApiVersion);
if (actualApiVersion !== currentApiVersion) {
  throw new Error(
    `You have multiple incompatible versions of @tldraw/state in your app. Please deduplicate the package.`
  );
}
export {
  ArraySet,
  EMPTY_ARRAY,
  EffectScheduler,
  RESET_VALUE,
  UNINITIALIZED,
  atom,
  computed,
  getComputedInstance,
  isAtom,
  isSignal,
  isUninitialized,
  react,
  reactor,
  transact,
  transaction,
  unsafe__withoutCapture,
  whyAmIRunning,
  withDiff
};
